### ISSUE TITLE
Terraform apply crashes with Provider produced inconsistent final plan after upgrading from helm 1.1.1 to any of helm 1.2.x

### ISSUE THREAD

**Issue by mkozioro** [hashicorp/terraform-provider-helm#541]
### Terraform Version and Provider Version
Terraform v0.12.23 but upgrading to v0.12.28 also does not solve issue

### Provider Version
v1.2.2 but issue was also with v1.2.0 and v1.2.1

### Affected Resource(s)
- helm_release

### Terraform Configuration Files
Deployment fails when using dynamic sets, like
```hcl
  dynamic "set" {
    for_each = var.prometheus_sets
    content {
      name  = set.key
      value = set.value
    }
  }
```

Example with our prometheus deployment chart. 
```hcl
resource "helm_release" "prometheus" {
  depends_on          = [helm_release.thanos]
  count               = var.enable_prometheus ? 1 : 0
  name                = "prometheus"
  repository          = var.storm_repository["url"]
  repository_username = var.storm_repository["username"]
  repository_password = var.storm_repository["password"]
  chart               = "prometheus"
  version             = "11.1.0-storm"
  namespace           = kubernetes_namespace.monitoring.metadata[0].name
  timeout             = 3600
  values = [
    fileexists(var.prometheus_config) ? file(var.prometheus_config) : ""
  ]

  set {
    name  = "server.ingress.enabled"
    value = var.enable_prometheus_ingress
  }

  set {
    name  = "server.ingress.hosts[0]"
    value = local.prometheus_host
  }

  set {
    name  = var.enable_prometheus_tls ? "server.ingress.tls[0].secretName" : ""
    value = module.prometheus_ingress_secret.secret_name
  }

  set {
    name  = var.enable_prometheus_tls ? "server.ingress.tls[0].hosts[0]" : ""
    value = local.prometheus_host
  }

  set {
    name  = "alertmanager.ingress.enabled"
    value = var.enable_alertmanager_ingress
  }

  set {
    name  = "alertmanager.ingress.hosts[0]"
    value = local.alertmanager_host
  }

  set {
    name  = var.enable_alertmanager_tls ? "alertmanager.ingress.tls[0].secretName" : ""
    value = module.alertmanager_ingress_secret.secret_name
  }

  set {
    name  = var.enable_alertmanager_tls ? "alertmanager.ingress.hosts[0]" : ""
    value = local.alertmanager_host
  }

  dynamic "set" {
    for_each = var.prometheus_sets
    content {
      name  = set.key
      value = set.value
    }
  }
}
```

prometheus_sets is built like:
```hcl
  prometheus_sets           = merge(
    {
      "server.baseURL" = "https://${azurerm_traffic_manager_profile.prometheus_tm_profile.fqdn}"
      "server.global.external_labels.cluster_name" = local.cluster_name
      "server.global.external_labels.location" = var.location
      "server.global.external_labels.environment" = var.environment
      "alertmanager.enabled" = local.enable_alertmanager
      "alertmanager.baseURL" = "http://${azurerm_traffic_manager_profile.prometheus_alertmanager_tm_profile.fqdn}"
      "alertmanager.ingress.hosts[0]" = azurerm_traffic_manager_profile.prometheus_alertmanager_tm_profile.fqdn
      "alertmanagerFiles.alertmanager\\.yml.global.smtp_auth_password" = var.smtp_auth_password,
      "server.image.repository": "${var.image_registry}/prom/prometheus",
      "server.image.tag": var.versions.external.docker["prom/prometheus"]
    },
    jsondecode(file("${path.module}/profiles/${var.profile}/prometheus.json")),
    {
      for idx, value in yamldecode(file("${path.module}/config/prometheus.yaml")).serverFiles["alerting_rules.yml"].groups[0].rules:
      "serverFiles.alerting_rules\\.yml.groups[0].rules[${idx}].expr" => replace(
        replace(
          replace(
            replace(value.expr, "<CAPACITY_ALERT_THRESHOLD_PLACEHOLDER>", var.profile == "prod" ? 0.75 : 0.95),
          ",", "\\,"),
        "{", "\\{"),
      "[", "\\[")
    }
  )
```
After checking with terraform console prometheus_sets map, none of keys or values are null.
Helm provider fails in any case where prometheus_sets or our other chart *_sets map is produced with merge of values from json and values derived from other values. 

### Debug Output
```
2020/07/07 14:17:00 [TRACE] GRPCProvider: ApplyResourceChange
2020-07-07T14:17:00.547+0200 [DEBUG] plugin.terraform-provider-helm_v1.2.2_x4: 2020/07/07 14:17:00 [DEBUG] setting computed for "metadata" from ComputedKeys
2020-07-07T14:17:00.548+0200 [DEBUG] plugin.terraform-provider-helm_v1.2.2_x4: 2020/07/07 14:17:00 [DEBUG] Getting Config
2020-07-07T14:17:01.335+0200 [DEBUG] plugin.terraform-provider-helm_v1.2.2_x4: 2020/07/07 14:17:01 [DEBUG] Getting chart
2020/07/07 14:17:01 [WARN] Provider "registry.terraform.io/-/helm" produced an invalid plan for module.ref_arch.helm_release.prometheus[0], but we are tolerating it because it is using the legacy plugin S
DK.
    The following problems may be the cause of any confusing errors from downstream operations:
      - .wait: planned value cty.True does not match config value cty.NullVal(cty.Bool)
      - .cleanup_on_fail: planned value cty.False does not match config value cty.NullVal(cty.Bool)
      - .disable_webhooks: planned value cty.False does not match config value cty.NullVal(cty.Bool)
      - .reset_values: planned value cty.False does not match config value cty.NullVal(cty.Bool)
      - .render_subchart_notes: planned value cty.True does not match config value cty.NullVal(cty.Bool)
      - .atomic: planned value cty.False does not match config value cty.NullVal(cty.Bool)
      - .force_update: planned value cty.False does not match config value cty.NullVal(cty.Bool)
      - .recreate_pods: planned value cty.False does not match config value cty.NullVal(cty.Bool)
      - .skip_crds: planned value cty.False does not match config value cty.NullVal(cty.Bool)
      - .replace: planned value cty.False does not match config value cty.NullVal(cty.Bool)
      - .disable_crd_hooks: planned value cty.False does not match config value cty.NullVal(cty.Bool)
      - .verify: planned value cty.False does not match config value cty.NullVal(cty.Bool)
      - .create_namespace: planned value cty.False does not match config value cty.NullVal(cty.Bool)
      - .dependency_update: planned value cty.False does not match config value cty.NullVal(cty.Bool)
      - .disable_openapi_validation: planned value cty.False does not match config value cty.NullVal(cty.Bool)
      - .lint: planned value cty.False does not match config value cty.NullVal(cty.Bool)
      - .max_history: planned value cty.NumberIntVal(0) does not match config value cty.NullVal(cty.Number)
      - .reuse_values: planned value cty.False does not match config value cty.NullVal(cty.Bool)
2020/07/07 14:17:01 [TRACE] module.ref_arch: eval: *terraform.EvalCheckPlannedChange
2020/07/07 14:17:01 [TRACE] EvalCheckPlannedChange: Verifying that actual change (action Create) matches planned change (action Create)
2020/07/07 14:17:01 [ERROR] module.ref_arch: eval: *terraform.EvalCheckPlannedChange, err: Provider produced inconsistent final plan: When expanding the plan for module.ref_arch.helm_release.prometheus[0]
 to include new values learned so far during apply, provider "registry.terraform.io/-/helm" produced an invalid new value for .set: block set length changed from 9 to 24.

This is a bug in the provider, which should be reported in the provider's own issue tracker.
2020/07/07 14:17:01 [ERROR] module.ref_arch: eval: *terraform.EvalSequence, err: Provider produced inconsistent final plan: When expanding the plan for module.ref_arch.helm_release.prometheus[0] to includ
e new values learned so far during apply, provider "registry.terraform.io/-/helm" produced an invalid new value for .set: block set length changed from 9 to 24.

This is a bug in the provider, which should be reported in the provider's own issue tracker.
2020/07/07 14:17:01 [TRACE] [walkApply] Exiting eval tree: module.ref_arch.helm_release.prometheus[0]
2020/07/07 14:17:01 [TRACE] vertex "module.ref_arch.helm_release.prometheus[0]": visit complete
2020/07/07 14:17:01 [WARN] Provider "registry.terraform.io/-/helm" produced an invalid plan for helm_release.ingress, but we are tolerating it because it is using the legacy plugin SDK.
    The following problems may be the cause of any confusing errors from downstream operations:
      - .recreate_pods: planned value cty.False does not match config value cty.NullVal(cty.Bool)
      - .skip_crds: planned value cty.False does not match config value cty.NullVal(cty.Bool)
      - .replace: planned value cty.False does not match config value cty.NullVal(cty.Bool)
      - .disable_crd_hooks: planned value cty.False does not match config value cty.NullVal(cty.Bool)
      - .verify: planned value cty.False does not match config value cty.NullVal(cty.Bool)
      - .lint: planned value cty.False does not match config value cty.NullVal(cty.Bool)
      - .max_history: planned value cty.NumberIntVal(0) does not match config value cty.NullVal(cty.Number)
      - .create_namespace: planned value cty.False does not match config value cty.NullVal(cty.Bool)
      - .dependency_update: planned value cty.False does not match config value cty.NullVal(cty.Bool)
      - .disable_openapi_validation: planned value cty.False does not match config value cty.NullVal(cty.Bool)
      - .reuse_values: planned value cty.False does not match config value cty.NullVal(cty.Bool)
      - .wait: planned value cty.True does not match config value cty.NullVal(cty.Bool)
      - .cleanup_on_fail: planned value cty.False does not match config value cty.NullVal(cty.Bool)
      - .disable_webhooks: planned value cty.False does not match config value cty.NullVal(cty.Bool)
      - .reset_values: planned value cty.False does not match config value cty.NullVal(cty.Bool)
      - .render_subchart_notes: planned value cty.True does not match config value cty.NullVal(cty.Bool)
      - .atomic: planned value cty.False does not match config value cty.NullVal(cty.Bool)
      - .force_update: planned value cty.False does not match config value cty.NullVal(cty.Bool)
2020/07/07 14:17:01 [TRACE] <root>: eval: *terraform.EvalCheckPlannedChange
2020/07/07 14:17:01 [TRACE] EvalCheckPlannedChange: Verifying that actual change (action Create) matches planned change (action Create)
2020/07/07 14:17:01 [TRACE] <root>: eval: *terraform.EvalGetProvider
2020/07/07 14:17:01 [TRACE] <root>: eval: *terraform.EvalReadState
2020/07/07 14:17:01 [TRACE] EvalReadState: reading state for helm_release.ingress
2020/07/07 14:17:01 [TRACE] EvalReadState: no state present for helm_release.ingress
2020/07/07 14:17:01 [TRACE] <root>: eval: *terraform.EvalReduceDiff
2020/07/07 14:17:01 [TRACE] <root>: eval: *terraform.EvalIf
2020/07/07 14:17:01 [TRACE] <root>: eval: terraform.EvalNoop
2020/07/07 14:17:01 [TRACE] <root>: eval: *terraform.EvalApplyPre
2020/07/07 14:17:01 [TRACE] <root>: eval: *terraform.EvalApply
```

### Expected Behavior
Terraform apply should pass like when using helm 1.1.1 provider.

### Actual Behavior
terraform apply with deployment having dynamic "set" produces error when set values are not coming from statically filled map, but map that is for example merge map values coming from .json file and values derived from other variables:
```
Error: Provider produced inconsistent final plan
When expanding the plan for module.ref_arch.helm_release.prometheus[0] to
include new values learned so far during apply, provider
"registry.terraform.io/-/helm" produced an invalid new value for .set: block
set length changed from 9 to 25.
This is a bug in the provider, which should be reported in the provider's own
issue tracker.
``` 
and fails. Running terraform apply second time works properly. 

### Steps to Reproduce

1. `terraform apply`

### References

- GH-1234
I've found similar errors in other providers:
https://github.com/hashicorp/terraform/issues/22048
https://github.com/terraform-providers/terraform-provider-aws/issues/7987
https://github.com/terraform-aws-modules/terraform-aws-eks/issues/527

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by camlow325** [comment]
fyi, there is some discussion in https://github.com/hashicorp/terraform/issues/25600#issuecomment-712288313 about the possibility of doing some work in Terraform core to avoid this sort of error.

**Comment by camlow325** [comment]
With a fix merged to the `master` branch in Terraform core, https://github.com/hashicorp/terraform/pull/26638, I'm no longer able to reproduce the "Provider produced inconsistent final plan" error when a `helm_release` resource generates `set`s via a `dynamic` block. Not sure when/if this would be backported to older Terraform versions.

**Comment by camlow325** [comment]
fyi, the fix for the "Provider produced inconsistent plan" issue regarding the use of a `dynamic` block with sets was included in the Terraform core [0.14.0-rc1 release](https://github.com/hashicorp/terraform/blob/v0.14.0-rc1/CHANGELOG.md):

> core: Prevent "Inconsistent Plan" errors when using dynamic with a block of TypeSet ([#26638](https://github.com/hashicorp/terraform/issues/26638))

**Comment by mkozioro** [comment]
@camlow325, Thanks for information.

**Comment by jrhouston** [comment]
Can anyone confirm if this is still an issue in the latest version of the provider?

**Comment by mkozioro** [comment]
Hi @jrhouston. It was fixed by upgrading terraform to 0.14 or 0.13.5 version and is not an issue any more.
